odoo/models.py

```diff
@@ -5590,35 +5590,68 @@ def exists(self):
         valid_ids = set([r[0] for r in self._cr.fetchall()] + new_ids)
         return self.browse(i for i in self._ids if i in valid_ids)
 
-    def _check_recursion(self, parent=None):
+    def _check_recursion(self, field_name=None):
```

Comment:
While we're at it, maybe we could rename this method.  The verb "check" feels inappropriate to me, since most "check" methods don't return anything, but raise some exception when something wrong is found.  The method actually looks for the presence of a cycle (or loop) from the records `self`.  What about something like `records._has_cycle(fname)` instead?

Reply:
@HydrionBurst @xmo-odoo

Reply:
Should I keep and deprecate the old names or just remove them?
Maybe `_has_cycle_in_relation` ("has cycle" seems quite short). I would keep file_name because it's more explicit, but I see both used in other methods.

Reply:
Maybe directly raise an error in the method? Since all uses cases are for raising ValidationError, we can directly raise error here with more information.
e.g. `SELECT` all record ids in the loop and in `self` instead of `LIMIT 1` and raise error with their `display_name`/`name`/`id`

but `records._has_cycle(fname)` also makes sense to me and seems better than `_check_recursion` if it returns bool

Reply:
Anyway it doesn't check for "recursion", so the word "recursion" was badly chosen as well.
I was also thinking to making the method raise an exception, but I wonder if people would agree on some generic error message.  The issue is that the error message is usually formulated in more functional terms, instead of purely technical terms.

Reply:
Indeed, looking at usages, the wording in the exception is always different. @rco-odoo What signature do you want for that method?

Reply:
I think you can just use `_has_cycle`. And since you change the method name, you can directly drop the `_check_m2m_recursion`

Reply:
Ok, pushed

Reply:
Given the comment below by @Feyensv  `_has_cycle` should return True if there is a cycle, so back to naming the function: `_is_acyclic`?

Note that since the function is renamed, enterprise code does not work anymore.

Reply:
I kept `_has_cycle` and added back old methods as deprecated to avoid breaking changes.